dataset/dataset.cpp

```diff
@@ -157,24 +137,38 @@ Dataset Dataset::slice(const Slice s) const {
   Dataset out;
   out.m_coords = m_coords.slice(s);
   out.m_data = slice_map(m_coords.sizes(), m_data, s);
-  // TODO dropping items independent of s.dim(), is this still what we want?
-  for (auto it = out.m_data.begin(); it != out.m_data.end();) {
-    if (!m_data.at(it->first).dims().contains(s.dim()))
-      it = out.m_data.erase(it);
-    else
-      ++it;
-  }
   for (auto it = m_coords.begin(); it != m_coords.end();) {
     if (unaligned_by_dim_slice(*it, s)) {
       auto extracted = out.m_coords.extract(it->first);
       for (auto &item : out.m_data)
-        item.second.attrs().set(it->first, extracted);
+        if (m_coords.item_applies_to(it->first, m_data.at(item.first).dims()))
+          item.second.attrs().set(it->first, extracted, true);
     }
     ++it;
   }
   return out;
 }
 
+Dataset &Dataset::setSlice(const Slice s, const Dataset &data) {
+  // TODO Need dry-run mechanism here?
+  for (const auto &item : data)
+    (*this)[item.name()].setSlice(s, item);
+  return *this;
+}
+
+Dataset &Dataset::setSlice(const Slice s, const DataArray &data) {
+  // TODO Need dry-run mechanism here?
```

Comment:
I added https://github.com/scipp/scipp/projects/9#card-59105799 to we don't forget. Suggest to not do this in this PR since it would take even longer.